### PR TITLE
Fix grouped actor errors for preconcurrency

### PR DESCRIPTION
**Pull request by angela-laar**
The experimental feature that groups actor isolation errors should downgrade to a warning if `@preconcurrency` attribute is applied.

### CODE REVIEW

**Comment by angela-laar** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeCheckConcurrency.cpp [hunk 1]
```diff
@@ -2116,21 +2116,25 @@ namespace {
 
         auto behavior = DiagnosticBehavior::Error;
 
+        // Limit behavior if @preconcurrency detected
+        if (errors.front().preconcurrency)
```
Comment (hborla) [inline lib/Sema/TypeCheckConcurrency.cpp:2120]: We can't just look at the first error here. We'll need to check that all of the errors are `.preconcurrency` before we can downgrade the error to a warning.

**Comment by angela-laar** [comment]
@swift-ci please smoke test

**Comment by angela-laar** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeCheckConcurrency.cpp [hunk 2]
```diff
@@ -2139,23 +2145,31 @@ namespace {
...
         }
 
         for (IsolationError error : errors) {
           // Diagnose actor_isolated_call as note if
-          // fix-it provided in missingGlobalActorOnContext
-          ctx.Diags.diagnose(error.loc, error.diag)
-              .limitBehaviorUntilSwiftVersion(behavior, 6);
+          // if there are multiple actor-isolated function calls
+          // from outside the actor
+            ctx.Diags.diagnose(error.loc, error.diag)
```
Comment (hborla) [inline lib/Sema/TypeCheckConcurrency.cpp:2171]: Nit: fix the indentation here; this line should be aligned with the comment above (with the `.limitBehavior` line below indented)

File: lib/Sema/TypeCheckConcurrency.cpp [hunk 3]
```diff
@@ -3375,15 +3389,19 @@ namespace {
 
           IsolationError mismatch([calleeDecl, apply, unsatisfiedIsolation, getContextIsolation]() {
             if (calleeDecl) {
+              auto preconcurrency = getContextIsolation().preconcurrency() || getActorIsolation(calleeDecl).preconcurrency();
```
Comment (hborla) [inline lib/Sema/TypeCheckConcurrency.cpp:3392]: Nit: please wrap this line

File: lib/Sema/TypeCheckConcurrency.cpp [hunk 4]
```diff
@@ -3401,8 +3419,8 @@ namespace {
           }
         } else {
           if (calleeDecl) {
-            auto preconcurrency = getContextIsolation().preconcurrency() ||
-                getActorIsolation(calleeDecl).preconcurrency();
+            auto preconcurrency = getContextIsolation().preconcurrency() || getActorIsolation(calleeDecl).preconcurrency();
```
Comment (hborla) [inline lib/Sema/TypeCheckConcurrency.cpp:3422]: Nit: please wrap this line

File: test/Concurrency/grouped_actor_isolation_diagnostics.swift [hunk 5]
```diff
@@ -81,3 +111,15 @@ struct HasWrapperOnActor {
...
+@preconcurrency @MainActor
+class MainActorPreconcurrency {}
+
+class InferMainActorPreconcurrency: MainActorPreconcurrency {
+  static func predatesConcurrency() {}
+}
+
+nonisolated func testPreconcurrency() {
+  InferMainActorPreconcurrency.predatesConcurrency()
+  // expected-warning@-1 {{call to main actor-isolated static method 'predatesConcurrency()' in a synchronous nonisolated context}}
```
Comment (hborla) [inline test/Concurrency/grouped_actor_isolation_diagnostics.swift:124]: Can you please add a test that has multiple callsites to `@preconcurrency` functions to ensure the diagnostic is still downgraded in that case?

**Review by hborla — APPROVED** [review]
LGTM, I left a few minor formatting nits and a request for another test case

File: lib/Sema/TypeCheckConcurrency.cpp [hunk 6]
```diff
@@ -4926,15 +4947,15 @@ ActorIsolation ActorIsolationRequest::evaluate(
     return isolation;
   };
 
-  auto isolationFromAttr = getIsolationFromAttributes(value);
+  auto isolationFromAttr = getIsolationFromAttributes(value);      // add to interface request
```
Comment (ktoso) [inline lib/Sema/TypeCheckConcurrency.cpp:4950]: somewhat weird comments, remove the "floaty" comments or move them above the line of code it is about?

**Review by ktoso — APPROVED** [review]

File: lib/Sema/TypeCheckConcurrency.cpp (hunk 6 above)
Comment (angela-laar) [reply lib/Sema/TypeCheckConcurrency.cpp:4950]: oh no, looks like this comment slipped in from another branch , will remove it!

**Review by ktoso — APPROVED** [review]
LGTM now, thanks for fixing the comments :)

**Comment by angela-laar** [comment]
@swift-ci please smoke test

**Pull request merged**